### PR TITLE
Enable dynamic resolution input for Beit

### PR DESCRIPTION
**Pull request by OmarManzoor**
# What does this PR do?

Towards #30579

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
CC: @amyeroberts 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by OmarManzoor** [comment]
Hi @amyeroberts 

This PR is incomplete right now because I am unsure how to proceed. It seems that the BeitEncoder takes the original patch embeddings configuration as an input and hence the original window size
https://github.com/huggingface/transformers/pull/31053/files#diff-3f84bebd6be8d9c0f5c5068199f5c49eac8489d5fa466fb6fa08b0365e78dba4R679-R685
This is later used to initialize the `BeitRelativePositionBias` and the dimensions mismatch. However we only interpolate the embeddings during the forward call. Could you kindly guide what I am missing?

File: src/transformers/models/beit/modeling_beit.py [hunk 1]
```diff
@@ -191,13 +239,25 @@ def __init__(self, config):
...
+        position_embedding: Optional[torch.Tensor] = None,
+        interpolate_pos_encoding: bool = False,
+    ) -> torch.Tensor:
         batch_size, num_channels, height, width = pixel_values.shape
         if num_channels != self.num_channels:
             raise ValueError(
                 "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
             )
 
+        if not interpolate_pos_encoding:
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:254]: No need for this in the patch embeddings

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! Just a few small comments

**Comment by OmarManzoor** [comment]
> @OmarManzoor Regarding the relative position bias, looking at the modeling code, I think whenever the output of this model is used, then it will also need to be interpolated if `interpolate_pos_encoding=True`

Could you kindly clarify a bit where exactly this should be added? Do we need to add a new interpolation function that works for BeitEncoder?

**Comment by amyeroberts** [comment]
@OmarManzoor You need to make sure that the relative position biases are interpolated wherever that is needed. This might be as an argument to the relative position class, or within the modules that use its output

**Comment by OmarManzoor** [comment]
> @OmarManzoor You need to make sure that the relative position biases are interpolated wherever that is needed.
For the BeitEmbeddings we used the embeddings for the interpolation. For BeitRelativePositionBias what should we use?

**Comment by OmarManzoor** [comment]
@amyeroberts How do I calculate the interpolations for the relative position biases similar to how we calculated them for the embeddings?

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 2]
```diff
@@ -670,6 +753,7 @@ def forward(
         head_mask: Optional[torch.Tensor] = None,
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
+        interpolate_pos_encoding: bool = False,
```
Comment (amyeroberts) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:756]: Should be added to `DATA2VEC_VISION_INPUTS_DOCSTRING`

**Review by amyeroberts — COMMENTED** [review]
Looks great - thanks for adding this! 

Only a small nit on the docstring for data2vec

**Review by amyeroberts — APPROVED** [review]

**Pull request merged**